test/scripts/run_test.py

```diff
@@ -33,42 +44,54 @@ def analyse_result(result):
-        logcat_cmd = ['logcat', '-d', target_package]
-        with open(logcat_file, 'a+') as lf:
-            subprocess.call(adb + logcat_cmd, stdout=lf)
+def analyse_unit_test_result(result):
+    total_num = 0
+    fail_num = 0
+    with open(result, 'r') as f:
+        for line in f:
+            m_failure = re.match('Tests run:\s+(\d+).*?Failures:\s+(\d+)', line)
+            m_succeed = re.match('OK.*?(\d+) test', line)
+            if m_failure is not None:
+                total_num = int(m_failure.group(1))
+                fail_num = int(m_failure.group(2))
+            if m_succeed is not None:
+                total_num = int(m_succeed.group(1))
+    return total_num, fail_num
+
+
+def exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file):
+    adb = ['adb'] if device == None else ['adb', '-s', device]
+    clean_logcat = ['logcat', '-c']
+    subprocess.call(adb + clean_logcat)
+    with open(result_file, 'a+') as rf:
+        subprocess.call(adb + am_cmd, stdout=rf)
+    logcat_cmd = ['logcat', '-d', target_package]
+    with open(logcat_file, 'a+') as lf:
+        subprocess.call(adb + logcat_cmd, stdout=lf)
+
+
+def run_cases(mode, module, target_package, log_dir, device, cases=None):
+    print '\n> running ' + mode.value + ' cases on device', device
+    result_file = os.path.join(log_dir, module + '-' + mode.value + '-result-' + LOGCAT_SUFFIX)
+    logcat_file = os.path.join(log_dir, module + '-' + mode.value + '-logcat-' + LOGCAT_SUFFIX)
+
+    if mode == TestMode.INSTRUMENTATION:
+        for case in cases:
+            am_cmd = ['shell', 'am', 'instrument', '-w', '-r', '-e', 'debug', 'false', '-e',
+                      'class', target_package + '.' + case,
+                      target_package + '.test/android.test.InstrumentationTestRunner']
+            exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file)
+    elif mode == TestMode.UNIT:
+        am_cmd = ['shell', 'am', 'instrument', '-w', '-r', '-e', 'debug', 'false',
+                  target_package + '.test/android.support.test.runner.AndroidJUnitRunner']
+        exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file)
     print '> done.'
-    print '  Result file: <LOG_DIR>/' + module + '-result-' + LOGCAT_SUFFIX
-    print '  Log file: <LOG_DIR>/' + module + '-logcat-' + LOGCAT_SUFFIX
-    return analyse_result(result_file)
-
-
-def install_test(module, device):
-    print '\n> building and installing test module', module
-    if module == 'base':
-        test_path = BASE_TEST_PATH
-    elif module == 'conference':
-        test_path = CONF_TEST_PATH
-    elif module == 'p2p':
-        test_path = P2P_TEST_PATH
+    print '  Result file: <LOG_DIR>/' + module + '-' + mode.value + '-result-' + LOGCAT_SUFFIX
+    print '  Log file: <LOG_DIR>/' + module + '-' + mode.value + '-logcat-' + LOGCAT_SUFFIX
+    return result_file
+
+
+def install_test(test_path, device):
```

Comment:
`device` is unused. and installDebug seems to install on all devices

Reply:
done

Reply:
Please test your code before submitting. This will have some issues in some situation.

Reply:
Its unnecessary to check this.
You don't have a clear understanding of your own code. When `device` is `None`, it means you don't specify a device, it should work on all devices.
And checking the device using `adb devices` is unnecessary either, it will fail upon install anyway.

Reply:
done